In [20]:
import pandas as pd

In [21]:
df=pd.read_csv('laptop_data.csv')
df.head()

Company   TypeName  Ram  Weight         Price  TouchScreen  Ips         Ppi  \
0   Apple  Ultrabook    8    1.37  11175.754549            0    1  226.983005   
1   Apple  Ultrabook    8    1.34  10776.777318            0    0  127.677940   
2      HP   Notebook    8    1.86  10329.931067            0    0  141.211998   
3   Apple  Ultrabook   16    1.83  11814.475945            0    1  220.534624   
4   Apple  Ultrabook    8    1.37  11473.100973            0    1  226.983005   

       Cpu_brand  HDD  SSD Gpu_brand      Os  
0  Intel Core i5    0  128     Intel     Mac  
1  Intel Core i5    0    0     Intel     Mac  
2  Intel Core i5    0  256     Intel  Others  
3  Intel Core i7    0  512       AMD     Mac  
4  Intel Core i5    0  256     Intel     Mac

In [22]:
df.dtypes

Company         object
TypeName        object
Ram              int64
Weight         float64
Price          float64
TouchScreen      int64
Ips              int64
Ppi            float64
Cpu_brand       object
HDD              int64
SSD              int64
Gpu_brand       object
Os              object
dtype: object

In [23]:
df.isnull().sum()

Company        0
TypeName       0
Ram            0
Weight         0
Price          0
TouchScreen    0
Ips            0
Ppi            0
Cpu_brand      0
HDD            0
SSD            0
Gpu_brand      0
Os             0
dtype: int64

In [24]:
df['Company'].value_counts().index


Index(['Dell', 'Lenovo', 'HP', 'Asus', 'Acer', 'MSI', 'Toshiba', 'Apple',
       'Samsung', 'Razer', 'Mediacom', 'Microsoft', 'Xiaomi', 'Vero', 'Chuwi',
       'Google', 'Fujitsu', 'LG', 'Huawei'],
      dtype='object', name='Company')

In [25]:
df['Company']=df['Company'].map({'Dell':0, 'Lenovo':1, 'HP':2, 'Asus':3, 'Acer':4, 'MSI':5, 'Toshiba':6, 'Apple':7,'Samsung':8, 'Razer':9, 'Mediacom':10, 'Microsoft':11, 'Xiaomi':12, 'Vero':13, 'Chuwi':14,'Google':15, 'Fujitsu':16, 'LG':17, 'Huawei':18})
df['Company']

0       7
1       7
2       2
3       7
4       7
       ..
1268    3
1269    1
1270    1
1271    1
1272    2
Name: Company, Length: 1273, dtype: int64

In [27]:
df['TypeName'].value_counts().index

Index(['Notebook', 'Gaming', 'Ultrabook', '2 in 1 Convertible', 'Workstation',
       'Netbook'],
      dtype='object', name='TypeName')

In [28]:
df['TypeName']=df['TypeName'].map({'Notebook':0, 'Gaming':1, 'Ultrabook':2, '2 in 1 Convertible':3, 'Workstation':4,'Netbook':5})
df['TypeName']

0       2
1       2
2       0
3       2
4       2
       ..
1268    0
1269    3
1270    3
1271    0
1272    0
Name: TypeName, Length: 1273, dtype: int64

In [29]:
df.dtypes

Company          int64
TypeName         int64
Ram              int64
Weight         float64
Price          float64
TouchScreen      int64
Ips              int64
Ppi            float64
Cpu_brand       object
HDD              int64
SSD              int64
Gpu_brand       object
Os              object
dtype: object

In [30]:
df['Gpu_brand'].value_counts().index

Index(['Intel', 'Nvidia', 'AMD'], dtype='object', name='Gpu_brand')

In [31]:
df['Gpu_brand']=df['Gpu_brand'].map({'Intel':0, 'Nvidia':1, 'AMD':2})
df['Gpu_brand']

0       0
1       0
2       0
3       2
4       0
       ..
1268    1
1269    0
1270    0
1271    0
1272    2
Name: Gpu_brand, Length: 1273, dtype: int64

In [32]:
df['Os'].value_counts().index

Index(['Windows', 'Others', 'Mac'], dtype='object', name='Os')

In [33]:
df['Os']=df['Os'].map({'Windows':0, 'Others':1, 'Mac':2})
df['Os']

0       2
1       2
2       1
3       2
4       2
       ..
1268    0
1269    0
1270    0
1271    0
1272    0
Name: Os, Length: 1273, dtype: int64

In [34]:
df.dtypes

Company          int64
TypeName         int64
Ram              int64
Weight         float64
Price          float64
TouchScreen      int64
Ips              int64
Ppi            float64
Cpu_brand       object
HDD              int64
SSD              int64
Gpu_brand        int64
Os               int64
dtype: object

In [35]:
df['Cpu_brand'].value_counts().index

Index(['Intel Core i7', 'Intel Core i5', 'Other Intel Processor',
       'Intel Core i3', 'AMD Processor'],
      dtype='object', name='Cpu_brand')

In [36]:
df['Cpu_brand']=df['Cpu_brand'].map({'Intel Core i7':0, 'Intel Core i5':1, 'Other Intel Processor':2,'Intel Core i3':3, 'AMD Processor':4})
df['Cpu_brand']

0       1
1       1
2       1
3       0
4       1
       ..
1268    0
1269    0
1270    0
1271    2
1272    0
Name: Cpu_brand, Length: 1273, dtype: int64

In [37]:
df

Company  TypeName  Ram  Weight         Price  TouchScreen  Ips  \
0           7         2    8    1.37  11175.754549            0    1   
1           7         2    8    1.34  10776.777318            0    0   
2           2         0    8    1.86  10329.931067            0    0   
3           7         2   16    1.83  11814.475945            0    1   
4           7         2    8    1.37  11473.100973            0    1   
...       ...       ...  ...     ...           ...          ...  ...   
1268        3         0    4    2.20  10555.256584            0    0   
1269        1         3    4    1.80  10433.899310            1    1   
1270        1         3   16    1.30  11288.114524            1    1   
1271        1         0    2    1.50   9409.283030            0    0   
1272        2         0    6    2.19  10614.128815            0    0   

             Ppi  Cpu_brand   HDD  SSD  Gpu_brand  Os  
0     226.983005          1     0  128          0   2  
1     127.677940          1     0    0          0   2  
2     141.211998          1     0  256          0   1  
3     220.534624          0     0  512          2   2  
4     226.983005          1     0  256          0   2  
...          ...        ...   ...  ...        ...  ..  
1268  100.454670          0   500    0          1   0  
1269  157.350512          0     0  128          0   0  
1270  276.053530          0     0  512          0   0  
1271  111.935204          2     0    0          0   0  
1272  100.454670          0  1000    0          2   0  

[1273 rows x 13 columns]

In [42]:
df['Price']=df['Price'].astype(int)
df['Price']

0       11175
1       10776
2       10329
3       11814
4       11473
        ...  
1268    10555
1269    10433
1270    11288
1271     9409
1272    10614
Name: Price, Length: 1273, dtype: int32

In [49]:
y=df[['Price']]
y

Price
0     11175
1     10776
2     10329
3     11814
4     11473
...     ...
1268  10555
1269  10433
1270  11288
1271   9409
1272  10614

[1273 rows x 1 columns]

In [46]:
x=df.drop('Price',axis=1)

In [47]:
x

Company  TypeName  Ram  Weight  TouchScreen  Ips         Ppi  Cpu_brand  \
0           7         2    8    1.37            0    1  226.983005          1   
1           7         2    8    1.34            0    0  127.677940          1   
2           2         0    8    1.86            0    0  141.211998          1   
3           7         2   16    1.83            0    1  220.534624          0   
4           7         2    8    1.37            0    1  226.983005          1   
...       ...       ...  ...     ...          ...  ...         ...        ...   
1268        3         0    4    2.20            0    0  100.454670          0   
1269        1         3    4    1.80            1    1  157.350512          0   
1270        1         3   16    1.30            1    1  276.053530          0   
1271        1         0    2    1.50            0    0  111.935204          2   
1272        2         0    6    2.19            0    0  100.454670          0   

       HDD  SSD  Gpu_brand  Os  
0        0  128          0   2  
1        0    0          0   2  
2        0  256          0   1  
3        0  512          2   2  
4        0  256          0   2  
...    ...  ...        ...  ..  
1268   500    0          1   0  
1269     0  128          0   0  
1270     0  512          0   0  
1271     0    0          0   0  
1272  1000    0          2   0  

[1273 rows x 12 columns]

In [50]:
x.shape,y.shape

((1273, 12), (1273, 1))

## BY LINEAR REGRESSION

In [51]:
from sklearn.model_selection import train_test_split

In [52]:
x_train , x_test, y_train,y_test=train_test_split(x,y,test_size=0.2)

In [53]:
from sklearn.linear_model import LinearRegression

In [54]:
lr = LinearRegression()

In [55]:
lr.fit(x_train,y_train)

LinearRegression()

In [56]:
lr.score(x_train,y_train)

0.7186384993785252

In [62]:
round(lr.score(x_train,y_train)*100)

72

ACCURACY IS 72 WHICH IS VERY LOW

In [57]:
pred=lr.predict(x_test)

In [58]:
y_test['prediction'] = pred

In [59]:
y_test.columns=['actual','prediction']

CHECKING THE PRICE MANUALLY

In [60]:
y_test

actual    prediction
1268   10555  10596.312449
892    11413  11088.879439
1167   10124  10189.877261
1158   10645  10688.026030
267    10121  10367.033767
...      ...           ...
883    10481  10885.421067
1108    9962  10013.052463
538    11470  11028.299189
120    10147  10124.681311
1226   10232  10236.224155

[255 rows x 2 columns]

In [61]:
y_test.iloc[0:20]

actual    prediction
1268   10555  10596.312449
892    11413  11088.879439
1167   10124  10189.877261
1158   10645  10688.026030
267    10121  10367.033767
510    10776  10769.522238
1195    9883   9873.276593
91     10510  10842.840283
207    11550  11559.076620
719    11065  11025.715106
637     9493  10117.167458
819    10602  10392.616424
735    10633  10681.243158
786    10968  10841.629292
659    12029  12605.424724
100    10553  11007.670924
1193   11064  10978.307752
858    10450  10516.169642
399    11848  12119.469064
347    11358  11753.549028

CHECKING ACCURACY WITH SQUARED METHOD

In [63]:
from sklearn.metrics import r2_score

In [64]:
round(r2_score(y_test['actual'],pred)*100,2) 

73.9

ACCURACY INCREASED BY 1.9%

NOW IMPLEMENTING LINEAR REGRESSION WITH POLYNOMIAL FEATURE WITH SQUARED METHOD

In [65]:
from sklearn.preprocessing import PolynomialFeatures

In [66]:
poly = PolynomialFeatures(degree=2)

In [67]:
x_train_transformed = poly.fit_transform(x_train)

In [68]:
lr2 = LinearRegression()
lr2.fit(x_train_transformed,y_train)

LinearRegression()

In [69]:
x_test_transformed = poly.fit_transform(x_test)

In [71]:
pred2 = lr2.predict(x_test_transformed) #declaring pred2 variable to store prediction 

In [72]:
r2_score(y_test['actual'],pred2)

0.8103544703688043

In [73]:
round(r2_score(y_test['actual'],pred2)*100,2) 

81.04

now we can see that previous accuracy is 72 now accuracy is 81.04 hence accuracy increased by 9.04%

## now by decision tree

In [74]:
from sklearn.tree import DecisionTreeRegressor

In [75]:
dtr = DecisionTreeRegressor()

In [76]:
dtr.fit(x_train,y_train)

DecisionTreeRegressor()

In [78]:
dtr.score(x_train,y_train) ## we got the 99% accuracy with training data

0.996164594644063

In [80]:
dtr.score(x_test,y_test['actual'])

0.8140111745999965

In [81]:
round(dtr.score(x_test,y_test['actual'])*100)

81

now with test data we got the same accuracy as in linear regression with polynomial feature we can also say in case of decision tree that fit is a good fit